src/exercises/exerciseTypes/orderWords/OrderWords.js

```diff
@@ -0,0 +1,555 @@
+    // Aligns all the status to be the same.
+    let updatedWords = [...wordsMasterStatus];
+    let inUseIds = [...constructorWordArray].map(word => word.id);
+    console.log(inUseIds)
+    for (let i = 0; i < updatedWords.length; i++) {
+      if (inUseIds.includes(updatedWords[i].id)) {
+        updatedWords[i].status = status;
+      }
+    }
+    setWordsMasterStatus(updatedWords);
+
+    let newConstructorWordArray = [...constructorWordArray];
+    for (let i = 0; i < newConstructorWordArray.length; i++) {
+      if (inUseIds.includes(newConstructorWordArray[i].id)) {
+        newConstructorWordArray[i].status = status;
+      }
+    }
+    setConstructorWordArray(newConstructorWordArray);
+  }
+
+  function resetSwapWordStatus() {
+    setWordSwapStatus("");
+    setwordSwapId(-1);
+  }
+
+  function handleUndoSelection() {
+
+    let newConstructorWordArray = [...constructorWordArray];
+    let newWordMasterStatus = [...wordsMasterStatus];
+
+    let inOrderWord = getWordById(wordSwapId, newConstructorWordArray);
+    let wordMastStatus = getWordById(wordSwapId, newWordMasterStatus);
+
+    // Set to previous state
+    inOrderWord.status = wordSwapStatus;
+    wordMastStatus.status = wordSwapStatus;
+
+    resetSwapWordStatus();
+    setConstructorWordArray(newConstructorWordArray);
+    setWordsMasterStatus(newWordMasterStatus);
+  }
+
+
+  function handleCheck() {
+    resetSwapWordStatus();
+    setHintCounter(hintCounter + 1);
+    // Do nothing if empty
+    if (constructorWordArray.length === 0) { return; };
+
+    // Check if the solution is already the same
+    let filterPunctuationOGText = removePunctuation(exerciseContext);
+
+    let constructedSentence = []
+    for (let i = 0; i < constructorWordArray.length; i++) {
+      constructedSentence.push(constructorWordArray[i].word);
+    }
+
+    // Get the Sentence
+    constructedSentence = constructedSentence.join(" ")
+    if (constructedSentence === filterPunctuationOGText) {
+      setHasClues(false);
+      setIsCorrect(true);
+      setAllInWordsStatus("correct");
+      let concatMessage = messageToAPI + "C";
+      handleAnswer(concatMessage);
+      // Need to reset MasterStatus, so when the new exercise is loaded, there are no words.
+      // This happens when you don't change the exercise.
+      setWordsMasterStatus([]);
+    }
+    else {
+      // Handle the Errors and Set the feedback
+      // API should return the status, then update the MasterStatus
+      console.log("Sentence incorrect!");
+
+      // We need to ensure that we don't send the entire sentence,
+      // or alignment might align very distant words.
+      let resizeSol = "" + filterPunctuationOGText
+      resizeSol = resizeSol.split(" ")
+      resizeSol = resizeSol.slice(0, constructorWordArray.length + 1).join(" ")
+      api.annotateClues(constructorWordArray, resizeSol, exerciseLang, (updatedMasterStatus) => {
+        // Variable to update and store in the user Activity.
```

Comment:
This is too much code, and too complicated for me to review :) 
From the software engineering POV I'd say that it would be good if you could split this very large message into several separate functions. Although I'm not sure how or whether it really works :)

Reply:
Is this for the annotateClues call specifically or in general? 

These highlighted lines in specific, basically ensure that the solution sent is the size of the sentence constructed so far + 1. This is done mainly to avoid matching words that are very far apart, which would cause the feedback to be very confusing. All the code in the callback is to parse the information that comes from the API. Basically, the API updates the status of all words in the construction box by setting some properties in it. I then iterate through them to get the first clues and the total number of errors. This is then reported to the user activity data. I tried to move them into more self-contained functions, that are a bit more explicit on what they do.

Let me know if you'd rather have a small meeting to go over the code.